File: ppcbudget/perfjoin.py
"""Optional join against a campaign performance report.

The change-history export carries no per-campaign spend, so dollar figures are
limited to the ~9% of campaigns whose budget happens to appear in a budget
change row. Dropping in any Amazon Ads campaign report that has Campaign,
Spend, Sales and Budget columns lifts that to full coverage.

Join coverage is reported in both directions -- a silent join failure is how
dashboards start lying.
"""

from __future__ import annotations

import csv
import re
from dataclasses import dataclass, field
from pathlib import Path

import openpyxl

from .scoring import CampaignDay

# Header aliases seen across Amazon Ads report variants.
ALIASES = {
    "campaign": ["campaign", "campaign name", "campaigns"],
    "spend": ["spend", "cost", "total spend"],
    "sales": ["sales", "total sales", "14 day total sales", "7 day total sales",
              "attributed sales", "total advertising cost of sales"],
    "budget": ["budget", "daily budget", "campaign daily budget"],
    "impressions": ["impressions", "impr"],
    "clicks": ["clicks"],
    "orders": ["orders", "total orders", "14 day total orders", "7 day total orders"],
    "roas": ["roas", "total roas", "return on ad spend"],
}


@dataclass(slots=True)
class PerfRecord:
    campaign_raw: str
    spend: float | None = None
    sales: float | None = None
    budget: float | None = None
    impressions: float | None = None
    clicks: float | None = None
    orders: float | None = None
    roas: float | None = None


@dataclass(slots=True)
class JoinReport:
    path: Path
    rows_read: int = 0
    matched: int = 0
    unmatched_perf: list[str] = field(default_factory=list)
    unmatched_history: list[str] = field(default_factory=list)
    budgets_added: int = 0
    roas_added: int = 0

    @property
    def coverage(self) -> float:
        total = self.matched + len(self.unmatched_history)
        return self.matched / total if total else 0.0


def campaign_key(name: str) -> str:
    """Normalize for joining: casefold, collapse whitespace, strip punctuation runs."""
    return re.sub(r"\s+", " ", str(name).strip()).casefold()


def _num(value) -> float | None:
    if value is None or value == "":
        return None
    text = re.sub(r"[^\d.\-]", "", str(value))
    if text in ("", "-", ".", "-."):
        return None
    try:
        return float(text)
    except ValueError:
        return None


def _map_headers(header: list) -> dict[str, int]:
    lookup = {}
    normalized = [re.sub(r"\s+", " ", str(h or "").strip().lower()) for h in header]
    for field_name, names in ALIASES.items():
        for i, h in enumerate(normalized):
            if h in names:
                lookup[field_name] = i
                break
    return lookup


def _read_rows(path: Path) -> tuple[list, list[list]]:
    if path.suffix.lower() in (".csv", ".tsv", ".txt"):
        delim = "\t" if path.suffix.lower() == ".tsv" else ","
        with path.open(newline="", encoding="utf-8-sig") as fh:
            rows = list(csv.reader(fh, delimiter=delim))
        return (rows[0], rows[1:]) if rows else ([], [])

    wb = openpyxl.load_workbook(path, read_only=True, data_only=True)
    ws = wb[wb.sheetnames[0]]
    rows = list(ws.iter_rows(values_only=True))
    wb.close()
    if not rows:
        return [], []
    # Some Amazon reports carry a title block before the real header.
    for i, row in enumerate(rows[:10]):
        if row and _map_headers(list(row)).get("campaign") is not None:
            return list(row), [list(r) for r in rows[i + 1:]]
    return list(rows[0]), [list(r) for r in rows[1:]]


def load_performance(path: str | Path) -> tuple[dict[str, PerfRecord], JoinReport]:
    path = Path(path)
    header, rows = _read_rows(path)
    cols = _map_headers(header)
    if "campaign" not in cols:
        raise ValueError(
            f"{path.name}: no Campaign column found. Expected one of: "
            + ", ".join(ALIASES['campaign'])
        )

    report = JoinReport(path=path)
    records: dict[str, PerfRecord] = {}

    def at(row, name):
        i = cols.get(name)
        return row[i] if i is not None and i < len(row) else None

    for row in rows:
        if not row or not at(row, "campaign"):
            continue
        raw = str(at(row, "campaign")).strip()
        key = campaign_key(raw)
        report.rows_read += 1
        rec = records.get(key) or PerfRecord(campaign_raw=raw)
        for f in ("spend", "sales", "budget", "impressions", "clicks", "orders", "roas"):
            v = _num(at(row, f))
            if v is not None:
                prev = getattr(rec, f)
                # Reports can be split by day/placement; sum the additive ones.
                setattr(rec, f, v if prev is None or f in ("budget", "roas") else prev + v)
        records[key] = rec

    for rec in records.values():
        if rec.roas is None and rec.spend and rec.sales is not None and rec.spend > 0:
            rec.roas = rec.sales / rec.spend
    return records, report


def apply_to(days: list[CampaignDay], records: dict[str, PerfRecord],
             report: JoinReport) -> None:
    """Overlay observed budgets onto scored days; unmatched names are reported."""
    seen: set[str] = set()
    for day in days:
        key = campaign_key(day.campaign)
        rec = records.get(key)
        if rec is None:
            report.unmatched_history.append(day.campaign)
            continue
        seen.add(key)
        report.matched += 1
        if rec.budget is not None and rec.budget > 0:
            if day.budget.source == "unknown":
                report.budgets_added += 1
            day.budget.value = rec.budget
            day.budget.time_weighted = rec.budget
            day.budget.source = "perf_report"
        if rec.roas is not None:
            report.roas_added += 1
        day.perf = rec  # type: ignore[attr-defined]

    report.unmatched_perf = [r.campaign_raw for k, r in records.items() if k not in seen]

File: src/lib/perfjoin.ts
/**
 * Optional join against a campaign performance report.
 *
 * The change-history export carries no per-campaign spend, so dollar figures are
 * limited to the ~9% of campaigns whose budget happens to appear in a budget
 * change row. Dropping in any Amazon Ads campaign report that has Campaign,
 * Spend, Sales and Budget columns lifts that to full coverage.
 *
 * Join coverage is reported in both directions -- a silent join failure is how
 * dashboards start lying.
 */

import * as XLSX from 'xlsx';
import type { CampaignDay } from './scoring';

// Header aliases seen across Amazon Ads report variants.
const ALIASES: Record<string, string[]> = {
  campaign: ['campaign', 'campaign name', 'campaigns'],
  spend: ['spend', 'cost', 'total spend'],
  sales: [
    'sales',
    'total sales',
    '14 day total sales',
    '7 day total sales',
    'attributed sales',
    'total advertising cost of sales',
  ],
  budget: ['budget', 'daily budget', 'campaign daily budget'],
  impressions: ['impressions', 'impr'],
  clicks: ['clicks'],
  orders: ['orders', 'total orders', '14 day total orders', '7 day total orders'],
  roas: ['roas', 'total roas', 'return on ad spend'],
};

const NUMERIC_FIELDS = [
  'spend',
  'sales',
  'budget',
  'impressions',
  'clicks',
  'orders',
  'roas',
] as const;

type NumericField = (typeof NUMERIC_FIELDS)[number];

export type PerfRecord = { campaignRaw: string } & Record<NumericField, number | null>;

export interface JoinReport {
  fileName: string;
  rowsRead: number;
  matched: number;
  unmatchedPerf: string[];
  unmatchedHistory: string[];
  budgetsAdded: number;
  roasAdded: number;
}

export function coverage(r: JoinReport): number {
  const total = r.matched + r.unmatchedHistory.length;
  return total ? r.matched / total : 0;
}

/** Normalize for joining: casefold, collapse whitespace, strip punctuation runs. */
export function campaignKey(name: string): string {
  return String(name).trim().replace(/\s+/g, ' ').toLowerCase();
}

function num(value: unknown): number | null {
  if (value === null || value === undefined || value === '') return null;
  const text = String(value).replace(/[^\d.\-]/g, '');
  if (text === '' || text === '-' || text === '.' || text === '-.') return null;
  const v = Number(text);
  return Number.isFinite(v) ? v : null;
}

function mapHeaders(header: unknown[]): Map<string, number> {
  const lookup = new Map<string, number>();
  const normalized = header.map((h) =>
    String(h ?? '')
      .trim()
      .toLowerCase()
      .replace(/\s+/g, ' '),
  );
  for (const [field, names] of Object.entries(ALIASES)) {
    const i = normalized.findIndex((h) => names.includes(h));
    if (i !== -1) lookup.set(field, i);
  }
  return lookup;
}

/** RFC 4180 reader: quoted fields, embedded delimiters and newlines. */
function parseDelimited(text: string, delim: string): unknown[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = '';
  let quoted = false;
  let i = 0;
  const body = text.charCodeAt(0) === 0xfeff ? text.slice(1) : text; // strip the BOM

  const endField = () => {
    row.push(field);
    field = '';
  };
  const endRow = () => {
    endField();
    rows.push(row);
    row = [];
  };

  while (i < body.length) {
    const ch = body[i];
    if (quoted) {
      if (ch === '"') {
        if (body[i + 1] === '"') {
          field += '"';
          i += 2;
          continue;
        }
        quoted = false;
        i += 1;
        continue;
      }
      field += ch;
      i += 1;
    } else if (ch === '"' && field === '') {
      quoted = true;
      i += 1;
    } else if (ch === delim) {
      endField();
      i += 1;
    } else if (ch === '\r' || ch === '\n') {
      endRow();
      if (ch === '\r' && body[i + 1] === '\n') i += 1;
      i += 1;
    } else {
      field += ch;
      i += 1;
    }
  }
  if (field !== '' || row.length) endRow();
  return rows;
}

function readRows(fileName: string, data: ArrayBuffer): { header: unknown[]; rows: unknown[][] } {
  const ext = fileName.slice(fileName.lastIndexOf('.')).toLowerCase();
  if (ext === '.csv' || ext === '.tsv' || ext === '.txt') {
    const text = new TextDecoder('utf-8').decode(data);
    const rows = parseDelimited(text, ext === '.tsv' ? '\t' : ',');
    return rows.length ? { header: rows[0], rows: rows.slice(1) } : { header: [], rows: [] };
  }

  const wb = XLSX.read(data, { type: 'array', dense: true, cellDates: false });
  const ws = wb.Sheets[wb.SheetNames[0]];
  const rows = ws
    ? XLSX.utils.sheet_to_json<unknown[]>(ws, {
        header: 1,
        raw: true,
        defval: null,
        blankrows: true,
      })
    : [];
  if (!rows.length) return { header: [], rows: [] };
  // Some Amazon reports carry a title block before the real header.
  for (let i = 0; i < Math.min(10, rows.length); i++) {
    if (rows[i] && mapHeaders(rows[i]).has('campaign')) {
      return { header: rows[i], rows: rows.slice(i + 1) };
    }
  }
  return { header: rows[0], rows: rows.slice(1) };
}

export function loadPerformance(
  fileName: string,
  data: ArrayBuffer,
): { records: Map<string, PerfRecord>; report: JoinReport } {
  const { header, rows } = readRows(fileName, data);
  const cols = mapHeaders(header);
  if (!cols.has('campaign')) {
    throw new Error(
      `${fileName}: no Campaign column found. Expected one of: ${ALIASES.campaign.join(', ')}`,
    );
  }

  const report: JoinReport = {
    fileName,
    rowsRead: 0,
    matched: 0,
    unmatchedPerf: [],
    unmatchedHistory: [],
    budgetsAdded: 0,
    roasAdded: 0,
  };
  const records = new Map<string, PerfRecord>();

  const at = (row: unknown[], name: string): unknown => {
    const i = cols.get(name);
    return i !== undefined && i < row.length ? row[i] : null;
  };

  for (const row of rows) {
    if (!row || !at(row, 'campaign')) continue;
    const raw = String(at(row, 'campaign')).trim();
    const key = campaignKey(raw);
    report.rowsRead += 1;
    let rec = records.get(key);
    if (!rec) {
      rec = {
        campaignRaw: raw,
        spend: null,
        sales: null,
        budget: null,
        impressions: null,
        clicks: null,
        orders: null,
        roas: null,
      };
    }
    for (const f of NUMERIC_FIELDS) {
      const v = num(at(row, f));
      if (v === null) continue;
      const prev = rec[f];
      // Reports can be split by day/placement; sum the additive ones.
      rec[f] = prev === null || f === 'budget' || f === 'roas' ? v : prev + v;
    }
    records.set(key, rec);
  }

  for (const rec of records.values()) {
    if (rec.roas === null && rec.spend && rec.sales !== null && rec.spend > 0) {
      rec.roas = rec.sales / rec.spend;
    }
  }
  return { records, report };
}

/** Overlay observed budgets onto scored days; unmatched names are reported. */
export function applyTo(
  days: CampaignDay[],
  records: Map<string, PerfRecord>,
  report: JoinReport,
): void {
  const seen = new Set<string>();
  for (const day of days) {
    const key = campaignKey(day.campaign);
    const rec = records.get(key);
    if (!rec) {
      report.unmatchedHistory.push(day.campaign);
      continue;
    }
    seen.add(key);
    report.matched += 1;
    if (rec.budget !== null && rec.budget > 0) {
      if (day.budget.source === 'unknown') report.budgetsAdded += 1;
      day.budget.value = rec.budget;
      day.budget.timeWeighted = rec.budget;
      day.budget.source = 'perf_report';
    }
    if (rec.roas !== null) report.roasAdded += 1;
    day.perf = rec;
  }

  report.unmatchedPerf = [...records.entries()]
    .filter(([k]) => !seen.has(k))
    .map(([, r]) => r.campaignRaw);
}
